nexus/src/app/instance.rs

```diff
@@ -1357,6 +1384,127 @@ impl super::Nexus {
         Ok(())
     }
 
+    /// Compares the existing Propolis record in CRDB (`db_instance`) with a new
+    /// instance runtime state supplied from a sled agent, determines whether
+    /// any Propolis VMMs will be left unused as a result of this transition,
+    /// and releases the resources that were reserved to them.
+    async fn reclaim_unused_propolis_resources(
+        &self,
+        opctx: &OpContext,
+        new_runtime: &nexus::InstanceRuntimeState,
+        db_instance: &nexus_db_model::Instance,
+    ) -> Result<(), Error> {
+        let log = &self.log;
+        let instance_id = db_instance.id();
+
+        // Handle changes to the instance's active Propolis ID. The only
+        // expected change is from the current ID to a migration destination ID.
+        let old_propolis_id = db_instance.runtime().propolis_id;
+        if new_runtime.propolis_id != old_propolis_id {
+            match db_instance.runtime().dst_propolis_id {
+                Some(dst_id) if dst_id != new_runtime.propolis_id => {
+                    warn!(log,
+                          "active Propolis did not change to migration target";
+                          "instance_id" => %instance_id,
+                          "new_propolis_id" => %new_runtime.propolis_id,
+                          "old_propolis_id" => %old_propolis_id,
+                          "dst_propolis_id" => %dst_id);
+                }
+                None => {
+                    warn!(log,
+                          "active Propolis changed without an active migration";
```

Comment:
Not a blocker here but reminds me that maybe we should make `active_propolis_id` nullable on the instance record and allocate a new one whenever starting an instance. Anyways, that'd probably be part of making instance start a saga itself.

Reply:
Agreed--the active sled should also be nullable and should be allowed to change in that saga. We have #2315 tracking nullable sled IDs and #2824 for making instance start a saga; I think having a nullable Propolis ID will fall out pretty quickly when we break ground on #2824.